src/rpc/server/linux_io_uring.zig

```diff
@@ -0,0 +1,837 @@
+    const InitErrOrEmpty = if (!can_use) error{} else InitError;
+    const InitResultOrNoreturn = if (!can_use) noreturn else LinuxIoUring;
+    pub fn init(
+        /// Not stored, only used for some initial SQE preps, the pointer needn't remain stable.
+        server_ctx: *const server.Context,
+    ) InitErrOrEmpty!?InitResultOrNoreturn {
+        if (!can_use) return null;
+        var io_uring = IoUring.init(4096, 0) catch |err| return switch (err) {
+            error.SystemOutdated,
+            error.PermissionDenied,
+            => return null,
+            else => |e| e,
+        };
+        errdefer io_uring.deinit();
+
+        try prepMultishotAccept(&io_uring, server_ctx.tcp);
+        return .{ .io_uring = io_uring };
+    }
+
+    pub fn deinit(self: *LinuxIoUring) void {
+        self.io_uring.deinit();
+    }
+
+    pub const AcceptAndServeConnectionsError =
+        GetSqeRetryError ||
+        ConsumeOurCqeError ||
+        std.mem.Allocator.Error;
+
+    pub fn acceptAndServeConnections(
+        self: *LinuxIoUring,
+        server_ctx: *server.Context,
+    ) AcceptAndServeConnectionsError!void {
+        const timeout_ts: std.os.linux.kernel_timespec = comptime .{
+            .tv_sec = 1,
+            .tv_nsec = 0,
+        };
+
+        const timeout_sqe = try getSqeRetry(&self.io_uring);
+        timeout_sqe.prep_timeout(&timeout_ts, 1, 0);
+        timeout_sqe.user_data = 1;
+
+        _ = try self.io_uring.submit_and_wait(1);
+
+        var pending_cqes_buf: [255]std.os.linux.io_uring_cqe = undefined;
+        const pending_cqes_count = try self.io_uring.copy_cqes(&pending_cqes_buf, 0);
+        const cqes_pending = pending_cqes_buf[0..pending_cqes_count];
+
+        for (cqes_pending) |raw_cqe| {
+            // NOTE(ink): this is kind of hacky, should try refactoring this to use DOD-like indexes instead of pointers,
+            // that way we can allocate special static indexes instead of this.
+            if (raw_cqe.user_data == timeout_sqe.user_data) continue;
+            const our_cqe = OurCqe.fromCqe(raw_cqe);
+            try consumeOurCqe(self, server_ctx, our_cqe);
+        }
+    }
+};
+
+fn prepMultishotAccept(
+    io_uring: *IoUring,
+    tcp: std.net.Server,
+) GetSqeRetryError!void {
+    const sqe = try getSqeRetry(io_uring);
+    sqe.prep_multishot_accept(tcp.stream.handle, null, null, std.os.linux.SOCK.CLOEXEC);
+    sqe.user_data = @bitCast(Entry.ACCEPT);
+}
+
+const ConsumeOurCqeError =
+    HandleRecvBodyError ||
+    std.mem.Allocator.Error ||
+    connection.HandleAcceptError ||
+    connection.HandleRecvError ||
+    connection.HandleSendError ||
+    connection.HandleSpliceError;
+
+/// Panic message for handling `EAGAIN`; we're not using nonblocking sockets at all,
+/// so it should be impossible to receive that error, or for such an error to be
+/// triggered just from malicious connections.
+const EAGAIN_PANIC_MSG =
```

Comment:
global constants at top of file ?

Reply:
Why do you think they belong on top? What if it's only used by two functions in the file? Couldn't you put it next to those two functions?

Reply:
nbd but mainly for consistency - when i see a screaming case constant i auto just go to the top of the file to find it - searching between methods or wherever is just more work imo

Reply:
I could probably just inline this into the relevant portion of `consumeOurCqe` - originally was being used by a couple other functions, but ended up only being used in that.

Reply:
The advantage of putting it right next to where it's used is that you don't need to go anywhere. It's visible at the same time as its usage.

If it's not immediately visible, I use go-to-definition in the IDE (ctrl-click or F12 in vscode). I assume you use that for structs and functions?

Reply:
706f136b75d03abed6a66023acce4c00a76e85c2

Reply:
if its out of the editor's scope (ie, you will need to scroll for it) then you will still need to go somewhere - theres cases when your reading code not in the editor without go-to-def shortcuts